BoundaryMeshMapping.cpp

```diff
@@ -0,0 +1,279 @@
+        for (auto it = d_bdry_meshes[part]->elements_begin(); it != e_it_end; ++it)
+        {
+            Elem* e = *it;
+            // If this element has nodes on the bottom half of the cylinder, delete it.
+            for (unsigned int n = 0; n < e->n_nodes(); ++n)
+            {
+                if (e->node_ref(n)(1) < 0.0)
+                {
+                    d_bdry_meshes[part]->delete_elem(e);
+                    break;
+                }
+            }
+        }
+        auto n_it_end = d_bdry_meshes[part]->nodes_end();
+        for (auto it = d_bdry_meshes[part]->nodes_begin(); it != n_it_end; ++it)
+        {
+            Node* n = *it;
+            // If the node is on the bottom half of the cylinder, delete it.
+            if ((*n)(1) < 0.0) d_bdry_meshes[part]->delete_node(n);
+        }
+        d_bdry_meshes[part]->prepare_for_use();
+        d_bdry_meshes[part]->print_info();
+        d_bdry_eq_sys_vec[part] = std::move(libmesh_make_unique<EquationSystems>(*d_bdry_meshes[part]));
+        d_fe_data[part] = std::make_shared<FEData>(
+            d_object_name + "::FEData::" + std::to_string(part), *d_bdry_eq_sys_vec[part], true);
+
+        if (from_restart)
+        {
+            const std::string& file_name = get_libmesh_restart_file_name(
+                restart_read_dirname, d_object_name, restart_restore_number, part, d_libmesh_restart_file_extension);
+            const XdrMODE xdr_mode = (d_libmesh_restart_file_extension == "xdr" ? DECODE : READ);
+            const int read_mode =
+                EquationSystems::READ_HEADER | EquationSystems::READ_DATA | EquationSystems::READ_ADDITIONAL_DATA;
+            d_bdry_eq_sys_vec[part]->read(file_name, xdr_mode, read_mode, /*partition_agnostic*/ true);
+        }
+        else
+        {
+            auto& X_sys = d_bdry_eq_sys_vec[part]->add_system<ExplicitSystem>(d_coords_sys_name);
+            for (unsigned int d = 0; d < NDIM; ++d) X_sys.add_variable("X_" + std::to_string(d), FEType());
+            auto& dX_sys = d_bdry_eq_sys_vec[part]->add_system<ExplicitSystem>(d_disp_sys_name);
+            for (unsigned int d = 0; d < NDIM; ++d) dX_sys.add_variable("dX_" + std::to_string(d), FEType());
+            auto& W_sys = d_bdry_eq_sys_vec[part]->add_system<ExplicitSystem>(d_W_sys_name);
+            W_sys.add_variable("W", FEType());
+            W_sys.assemble_before_solve = false;
+            W_sys.assemble();
+        }
+    }
+
+    return;
+}
+
+void
+BoundaryMeshMapping::updateBoundaryLocation(const double t_start, const double t_end, const bool initial_time)
+{
+    plog << "Updating boundary location\n";
+    for (unsigned int part = 0; part < d_bdry_meshes.size(); ++part) updateBoundaryLocation(t_start, t_end, part);
+    if (initial_time)
+    {
+        setInitialConditions();
+    }
+    return;
+}
+
+void
+BoundaryMeshMapping::updateBoundaryLocation(const double t_start, const double t_end, const unsigned int part)
```

Comment:
This function makes sure the boundary mesh is always attached to the corresponding points on the volumetric mesh.

Note that `X_system` and `X_bdry_sys` are the volumetric and boundary mappings to their current configuration respecitvely (This would be chi in the IB literature).